nixpkgs/toolchains/go.bzl

```diff
@@ -7,31 +7,244 @@ dependencies on rules_go for those who don't need go toolchain.
+    """Copied from `rules_go`, since we have no other way to determine the proper `goarch` value.
+    """
+    if ctx.os.name == "linux":
+        goos, goarch = "linux", "amd64"
+        res = ctx.execute(["uname", "-p"])
+        if res.return_code == 0:
+            uname = res.stdout.strip()
+            if uname == "s390x":
+                goarch = "s390x"
+            elif uname == "i686":
+                goarch = "386"
+
+        # uname -p is not working on Aarch64 boards
+        # or for ppc64le on some distros
+        res = ctx.execute(["uname", "-m"])
+        if res.return_code == 0:
+            uname = res.stdout.strip()
+            if uname == "aarch64":
+                goarch = "arm64"
+            elif uname == "armv6l":
+                goarch = "arm"
+            elif uname == "armv7l":
+                goarch = "arm"
+            elif uname == "ppc64le":
+                goarch = "ppc64le"
+
+        # Default to amd64 when uname doesn't return a known value.
+
+    elif ctx.os.name == "mac os x":
+        goos, goarch = "darwin", "amd64"
+
+        res = ctx.execute(["uname", "-m"])
+        if res.return_code == 0:
+            uname = res.stdout.strip()
+            if uname == "arm64":
+                goarch = "arm64"
+
+        # Default to amd64 when uname doesn't return a known value.
+
+    elif ctx.os.name.startswith("windows"):
+        goos, goarch = "windows", "amd64"
+    elif ctx.os.name == "freebsd":
+        goos, goarch = "freebsd", "amd64"
+    else:
+        fail("Unsupported operating system: " + ctx.os.name)
+
+    return goos, goarch
+
+go_helpers_build = """
+load("@io_bazel_rules_go//go:def.bzl", "go_sdk")
+
+def go_sdk_for_arch():
+    native.filegroup(
+        name = "libs",
+        srcs = native.glob(
+            ["pkg/{goos}_{goarch}/**/*.a"],
+            exclude = ["pkg/{goos}_{goarch}/**/cmd/**"],
+        ),
+    )
+
+    go_sdk(
+        name = "go_sdk",
+        goos = "{goos}",
+        goarch = "{goarch}",
+        root_file = "ROOT",
+        package_list = ":package_list",
+        libs = [":libs"],
+        headers = [":headers"],
+        srcs = [":srcs"],
+        tools = [":tools"],
+        go = "bin/go{exe}",
+    )
+"""
+
+def _nixpkgs_go_helpers_impl(repository_ctx):
+    repository_ctx.file("BUILD.bazel", executable = False, content = "")
+    goos, goarch = _detect_host_platform(repository_ctx)
+    content = go_helpers_build.format(
+        goos = goos,
+        goarch = goarch,
+        exe = ".exe" if goos == "windows" else "",
+    )
+    repository_ctx.file("go_sdk.bzl", executable = False, content = content)
+
+nixpkgs_go_helpers = repository_rule(
+    implementation = _nixpkgs_go_helpers_impl
+)
+
+go_toolchain_func = """
+load("@io_bazel_rules_go//go/private:platforms.bzl", "PLATFORMS")
+load("@io_bazel_rules_go//go:def.bzl", "go_toolchain")
+
+def declare_toolchains(host_goos, host_goarch):
+    for p in [p for p in PLATFORMS if not p.cgo]:
```

Comment:
This is question coming completely out of ignorance: does this loop mean the cross-compilation capability is always prepared for all known golang PLATFORMS? Is that desired?